pkg/shared/utils/sync/distributed_lock.go

```diff
@@ -0,0 +1,61 @@
+package sync
+
+import (
+	"github.com/pkg/errors"
+	"gorm.io/gorm"
+)
+
+// DistributedLock manages locks on the database through the usage of the `distributed_lock` table.
+// WARNING: a lock can only be released by the instance that initially created it.
+type DistributedLock interface {
+	Lock() error
+	Unlock() error
+	isLocked() bool
+}
+
+var _ DistributedLock = &distributedLock{}
+
+// NewDistributedLock returns a new instance of DistributedLock
```

Comment:
Can we add more details on the behavior of this as explained in the description of this PR?